keps/0019-package-api-versioning.md

```diff
@@ -0,0 +1,72 @@
+---
+kep-number: 19
+title: Versioning of KUDO API in Operator Packages
+authors:
+  - "@nfnt"
+owners:
+  - "@nfnt"
+creation-date: 2019-11-05
+last-updated: 2019-11-05
+status: provisional
+---
+
+# Versioning of KUDO API in Operator Packages
+
+## Table of Contents
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+  * [Risks and Mitigations](#risks-and-mitigations)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+By adding a KUDO API version to package indexes, we will avoid that changes to the operator schema will affect existing KUDO versions and package definitions.
+
+## Motivation
+
+Operator packages are developed by defining the operator using YAML files. The schema to use for these files is defined by KUDO. While the KUDO team tries to keep the schema definitions consistent, there can be situations where the schema has to be changed. We need to ensure that packages are in sync with these changes while at the same time allowing existing packages to work with existing versions of KUDO.
+
+### Goals
+
+* Allow users to install packages from a package repository that provides packages for multiple KUDO API versions
+* Reject packages that don't have a matching KUDO API version
+
+### Non-Goals
+
+* Handle operator dependencies on specific Kubernetes versions
+
+## Proposal
+
+Through the `apiVersion` field in an `operator.yaml` it is already guaranteed that the operator will match a specific API version of KUDO. However, this field isn't taken into account when installing packages from a repository. KUDO will try to install the latest version of the package even if the `apiVersion` doesn't match. This can be mitigated by adding the packages `apiVersion` to the repository index and filtering out all packages that don't have a matching version. This way, older versions of packages can still be part of the repository and keep working with older versions of KUDO.
+
+For this, the following changes have to be made in KUDO:
+
+* Add `apiVersion` to the package index by updating the `Metadata` struct in `pkg/kudoctl/util/repo` and filling these values in `pkg/kudoctl/util/repo.ToPackageVersion`. A `index.yaml` entry should then look like this:
```

Comment:
what happens with the current kudoversion that is part of the operator.yaml? I think that should be removed. That should also be covered in this KEP I think

Reply:
also I think this should capture how will be package versioning look like from the operator developer perspective.

I have kafka version 2.0 that I release in operator version 1.1 which is API version v1beta1
New api version is added
I release kafka version 2.0 in operator version 2.1 with API version v1beta2

Can I still kind of update the v1beta1? If so, how?

Reply:
Sure, an operator can still get an update for `v1beta1`. Even today, without this KEP, because the `apiVersion` is referenced in `operators.yaml`. The only issue with this is the discoverability of this package through the repository index.

Reply:
This is interesting: The operator version could be the same, but the API version could change. Currently, only the operator version is used when naming the package tarball. There a three versions fully describing a package: The operator version, the API version and the app version. We should change the generated tarball name to reflect this. E.g. `kafka-0.1.2.tgz` should become `kafka-2.2.1-0.1.2-v1beta1.tgz`.

Reply:
Good point regarding `kudoVersion`, that could be removed. Or treated like `kubernetesVersion` in that it's an indicator of which versions of Kubernetes this operator was developed against without actually enforcing this version.

Reply:
Yeah I actually don't know what versions should be in the name and what it actually means, because that makes things complicated. Right now I install package with some version - I find that very straightforward. When I want to update, I go to package of higher version.

But do we e.g. allow kafka-1.0-v1beta1 and kafka-1.00-v1beta2? How do we structure that in the git repository? (current design is the simplest one when master has the latest - but what is the latest there? :)) Possibly we could even explore different options around that

Reply:
You raise a good question on how to manage that in git. If there's only the latest version of an operator in `master`, there shouldn't be any app version updates for older API versions. Of course, the repository will still have older package versions available, but updates will only be done for the latest API version and users need to use the respective version of KUDO if they want to install the latest app versions.